File: nodejs-example/named_channel_checkpoint.js
/**
 * Coordinate retained source offsets with one stable named channel.
 *
 * The offset token is checkpoint metadata, not a deduplication key. Give each
 * channel name one owner and retain events until committed progress is confirmed.
 */
"use strict";

const streaming = require("snowpipe-streaming");

const MAX_PENDING_EVENTS = 100_000;
const MAX_NO_PROGRESS_MS = 30 * 60_000;
const POLL_MS = 1_000;
const MAX_ATTEMPTS = 6;
const INVALIDATION_ERRORS = new Set([
  "InvalidChannelError",
  "InvalidClientError",
  "ClosedChannelError",
  "ClosedElasticChannelError",
  "ClosedClientError",
]);

async function createClient() {
  return streaming.createTableClient({
    clientName: `production-${process.pid}`,
    dbName: process.env.SNOWFLAKE_DATABASE || "MY_DATABASE",
    schemaName: process.env.SNOWFLAKE_SCHEMA || "MY_SCHEMA",
    tableName: process.env.SNOWFLAKE_TABLE || "MY_TABLE",
    profilePath: process.env.SNOWFLAKE_PROFILE || "profile.json",
  });
}

class SampleEventSource {
  constructor(total = 10_000, checkpoint = 0) {
    if (!Number.isSafeInteger(total) || checkpoint < 0 || checkpoint > total) {
      throw new Error("Require integer 0 <= checkpoint <= total");
    }
    this.total = total;
    this.committed = checkpoint;
    this.nextOffset = checkpoint + 1;
  }

  read() {
    if (this.nextOffset > this.total) return null;
    const offset = this.nextOffset++;
    return {
      offset,
      row: { EVENT_ID: offset, C1: offset, C2: `event-${offset}` },
    };
  }

  acknowledge(offset) {
    if (offset < this.committed || offset > this.total) {
      throw new Error("Invalid source checkpoint");
    }
    this.committed = offset;
  }

  seek(committed) {
    this.acknowledge(committed);
    this.nextOffset = committed + 1;
  }
}

function isInvalidation(error) {
  return INVALIDATION_ERRORS.has(error.errorCode);
}

function isRetryable(error) {
  return error instanceof streaming.StreamingIngestError
    && (isInvalidation(error)
      || [408, 429, 500, 502, 503, 504].includes(error.httpStatusCode));
}

function remaining(deadline) {
  const millis = deadline - performance.now();
  if (millis <= 0) {
    throw new Error("No confirmed progress before deadline; retain unconfirmed events");
  }
  return millis;
}

async function backoff(attempt, deadline) {
  const cap = Math.min(10_000, 250 * 2 ** Math.min(attempt, 6));
  const delay = Math.min(remaining(deadline), Math.random() * cap);
  await new Promise((resolve) => setTimeout(resolve, delay));
}

const CHANNEL_NAME = process.env.SNOWFLAKE_CHANNEL || "production-source-0";
const CHECKPOINT_ROWS = 1_000;
const CHECKPOINT_MS = 5_000;

function parseOffset(token) {
  if (token == null) return 0;
  const offset = Number(token);
  if (!Number.isSafeInteger(offset) || offset < 0) {
    throw new Error("Invalid committed source offset");
  }
  return offset;
}

class NamedProducer {
  constructor(factory = createClient) {
    this.factory = factory;
    this.client = null;
    this.channel = null;
  }

  async open() {
    if (!this.client) this.client = await this.factory();
    const opened = await this.client.openChannel({ name: CHANNEL_NAME });
    this.channel = opened.channel;
    if (opened.status.rowsErrorCount) {
      throw new Error("Row errors require reconciliation before source handoff");
    }
    return parseOffset(opened.status.latestCommittedOffsetToken);
  }

  async recover(error) {
    // Reopen without replacing the server-side committed offset.
    if (error.errorCode === "InvalidClientError") {
      await this.close(false);
    } else if (this.channel) {
      await this.channel.close({ waitForFlush: false, timeoutMs: 0 }).catch(() => {});
    }

    try {
      return await this.open();
    } catch (reopened) {
      if (!["InvalidClientError", "ClosedClientError"].includes(reopened.errorCode)) {
        throw reopened;
      }
      await this.close(false);
      return this.open();
    }
  }

  async close(flush) {
    if (!this.client) return;
    try {
      await this.client.close({ waitForFlush: flush, timeoutMs: 30_000 });
    } finally {
      this.client = null;
      this.channel = null;
    }
  }
}

async function collectProgress(producer, submitted, source) {
  const status = await producer.channel.getChannelStatus();
  if (status.rowsErrorCount) {
    throw new Error("Row errors require reconciliation before source handoff");
  }
  if (status.statusCode !== "SUCCESS") {
    throw new streaming.StreamingIngestError(
      "InvalidChannelError",
      status.statusCode,
      409,
      "Conflict",
    );
  }

  const committed = Math.min(
    submitted,
    parseOffset(status.latestCommittedOffsetToken),
  );
  if (committed > source.committed) source.acknowledge(committed);
}

async function run(producer, source) {
  // Snowflake's committed token determines where this retained source resumes.
  source.seek(await producer.open());
  let submitted = source.committed;
  let event = null;
  let exhausted = false;
  let failures = 0;
  let rowsSincePoll = 0;
  let nextPoll = performance.now() + CHECKPOINT_MS;
  let deadline = performance.now() + MAX_NO_PROGRESS_MS;

  while (true) {
    try {
      const outstanding = submitted > source.committed;
      const shouldPoll = outstanding
        && (exhausted
          || event
          || rowsSincePoll >= CHECKPOINT_ROWS
          || performance.now() >= nextPoll
          || submitted - source.committed >= MAX_PENDING_EVENTS);
      if (shouldPoll) {
        const previous = source.committed;
        await collectProgress(producer, submitted, source);
        if (source.committed > previous) {
          deadline = performance.now() + MAX_NO_PROGRESS_MS;
          failures = 0;
        }
        rowsSincePoll = 0;
        nextPoll = performance.now() + CHECKPOINT_MS;
      }

      if (submitted === source.committed && !event) {
        deadline = performance.now() + MAX_NO_PROGRESS_MS;
        if (exhausted) return;
      }

      remaining(deadline);
      if (exhausted
        || submitted - source.committed >= MAX_PENDING_EVENTS) {
        await new Promise((resolve) => {
          setTimeout(resolve, Math.min(POLL_MS, remaining(deadline)));
        });
        continue;
      }

      if (!event) event = source.read();
      if (!event) {
        exhausted = true;
        continue;
      }

      producer.channel.appendRow(event.row, String(event.offset));
      submitted = event.offset;
      rowsSincePoll++;
      event = null;
    } catch (error) {
      if (!isRetryable(error)) throw error;
      if (error.httpStatusCode !== 429 && ++failures >= MAX_ATTEMPTS) {
        throw error;
      }
      if (isInvalidation(error)) {
        const previous = source.committed;
        source.seek(await producer.recover(error));
        if (source.committed > previous) {
          deadline = performance.now() + MAX_NO_PROGRESS_MS;
        }
        submitted = source.committed;
        event = null;
        exhausted = false;
      }
      await backoff(2, deadline);
    }
  }
}

async function main() {
  const source = new SampleEventSource(
    Number(process.env.SNOWFLAKE_TEST_ROWS || 10_000),
  );
  const producer = new NamedProducer();
  let completed = false;
  try {
    await run(producer, source);
    completed = true;
    console.log(`Committed source checkpoint: ${source.committed}`);
  } finally {
    if (!completed) {
      console.error(`Retain source events after checkpoint ${source.committed}`);
    }
    await producer.close(completed);
  }
}

if (require.main === module) {
  const keepAlive = setInterval(() => {}, 1_000);
  main()
    .catch((error) => {
      console.error(error.message);
      process.exitCode = 1;
    })
    .finally(() => clearInterval(keepAlive));
}

module.exports = {
  MAX_ATTEMPTS,
  MAX_NO_PROGRESS_MS,
  MAX_PENDING_EVENTS,
  POLL_MS,
  SampleEventSource,
  NamedProducer,
  backoff,
  collectProgress,
  createClient,
  isInvalidation,
  isRetryable,
  main,
  parseOffset,
  remaining,
  run,
};

File: python-example/named_channel_checkpoint.py
"""Coordinate retained source offsets with a named channel.

This is an alternative to the Elastic progression for integrations that need
a stable channel identity and source-offset checkpointing.

A named channel reports its latest committed offset token when opened and in
channel status. The token is checkpoint metadata, not a deduplication key.
Give each stable channel name one owner and retain source events until its
committed progress is confirmed.
"""

import os
import random
import time
from dataclasses import dataclass
from typing import Optional

from snowflake.ingest import streaming

from elastic_step1_quickstart import connection_properties


MAX_ATTEMPTS = 6
MAX_NO_PROGRESS_SECONDS = 30 * 60.0
MAX_PENDING_EVENTS = 100_000
POLL_SECONDS = 1.0
INVALIDATION_ERRORS = {
    "InvalidChannelError",
    "InvalidClientError",
    "ClosedChannelError",
    "ClosedElasticChannelError",
    "ClosedClientError",
}
TRANSIENT_STATUS_CODES = {408, 429, 500, 502, 503, 504}


def create_client() -> streaming.StreamingIngestClient:
    properties = connection_properties()
    return streaming.StreamingIngestClient.from_table(
        client_name=f"recovery-{os.getpid()}",
        db_name=os.environ.get("SNOWFLAKE_DATABASE", "MY_DATABASE"),
        schema_name=os.environ.get("SNOWFLAKE_SCHEMA", "MY_SCHEMA"),
        table_name=os.environ.get("SNOWFLAKE_TABLE", "MY_TABLE"),
        profile_json=None if properties else os.environ.get("SNOWFLAKE_PROFILE", "profile.json"),
        properties=properties,
    )


@dataclass(frozen=True)
class Event:
    offset: int
    row: dict[str, object]


class SampleEventSource:
    """Regenerable sample data with an in-memory, non-durable checkpoint."""

    def __init__(self, total: int = 10_000, checkpoint: int = 0) -> None:
        if not 0 <= checkpoint <= total:
            raise ValueError("Require 0 <= checkpoint <= total")
        self.total = total
        self.committed = checkpoint
        self.next_offset = checkpoint + 1

    def read(self) -> Optional[Event]:
        if self.next_offset > self.total:
            return None
        offset = self.next_offset
        self.next_offset += 1
        return Event(
            offset,
            {"EVENT_ID": offset, "C1": offset, "C2": f"event-{offset}"},
        )

    def acknowledge(self, offset: int) -> None:
        if not self.committed <= offset <= self.total:
            raise ValueError("Invalid source checkpoint")
        self.committed = offset

    def seek(self, committed: int) -> None:
        self.acknowledge(committed)
        self.next_offset = committed + 1


def is_invalidation(error: streaming.StreamingIngestError) -> bool:
    return error.error_code.value in INVALIDATION_ERRORS


def is_retryable(error: BaseException) -> bool:
    return isinstance(error, streaming.StreamingIngestError) and (
        is_invalidation(error) or error.http_status_code in TRANSIENT_STATUS_CODES
    )


def remaining(deadline: float) -> float:
    seconds = deadline - time.monotonic()
    if seconds <= 0:
        raise TimeoutError(
            "No confirmed progress before the deadline; retain unconfirmed source events"
        )
    return seconds


def backoff(attempt: int, deadline: float) -> None:
    cap = min(10.0, 0.25 * 2 ** min(attempt, 6))
    time.sleep(min(random.uniform(0, cap), remaining(deadline)))


CHANNEL_NAME = os.environ.get("SNOWFLAKE_CHANNEL", "production-source-0")
CHECKPOINT_ROWS = 1_000
CHECKPOINT_SECONDS = 5.0


# Ingestion and status polling

def main() -> None:
    source = SampleEventSource(
        total=int(os.environ.get("SNOWFLAKE_TEST_ROWS", "10000"))
    )
    producer = NamedProducer()
    completed = False
    try:
        run(producer, source)
        completed = True
        print(f"Committed source checkpoint: {source.committed}")
    finally:
        if not completed:
            print(f"Retain source events after checkpoint {source.committed}")
        producer.close(completed)


def run(producer: "NamedProducer", source: SampleEventSource) -> None:
    """Append continuously and periodically checkpoint committed progress."""
    # Snowflake's committed token determines where this retained source resumes.
    source.seek(producer.open())
    submitted = source.committed
    event = None
    exhausted = False
    failures = 0
    rows_since_poll = 0
    next_poll = time.monotonic() + CHECKPOINT_SECONDS
    deadline = time.monotonic() + MAX_NO_PROGRESS_SECONDS

    while True:
        try:
            outstanding = submitted > source.committed
            should_poll = outstanding and (
                exhausted
                or event is not None
                or rows_since_poll >= CHECKPOINT_ROWS
                or time.monotonic() >= next_poll
                or submitted - source.committed >= MAX_PENDING_EVENTS
            )
            if should_poll:
                previous = source.committed
                # One status call can confirm a partial prefix; it does not
                # drain or wait for every submitted row.
                collect_progress(producer, submitted, source)
                if source.committed > previous:
                    deadline = time.monotonic() + MAX_NO_PROGRESS_SECONDS
                    failures = 0
                rows_since_poll = 0
                next_poll = time.monotonic() + CHECKPOINT_SECONDS

            if submitted == source.committed and event is None:
                deadline = time.monotonic() + MAX_NO_PROGRESS_SECONDS
                if exhausted:
                    return

            remaining(deadline)
            if exhausted or submitted - source.committed >= MAX_PENDING_EVENTS:
                time.sleep(min(POLL_SECONDS, remaining(deadline)))
                continue

            if event is None:
                event = source.read()
            if event is None:
                exhausted = True
                continue

            producer.channel.append_row(event.row, str(event.offset))
            submitted = event.offset
            rows_since_poll += 1
            event = None

        except streaming.StreamingIngestError as error:
            if not is_retryable(error):
                raise
            if error.http_status_code != 429:
                failures += 1
                if failures >= MAX_ATTEMPTS:
                    raise
            if is_invalidation(error):
                previous = source.committed
                source.seek(producer.recover(error))
                if source.committed > previous:
                    deadline = time.monotonic() + MAX_NO_PROGRESS_SECONDS
                submitted = source.committed
                event = None
                exhausted = False
            backoff(2, deadline)


# Committed offset handling

def collect_progress(producer: "NamedProducer", submitted: int, source: SampleEventSource) -> None:
    """Fetch status once and checkpoint the confirmed source prefix."""
    status = producer.channel.get_channel_status()
    if status.rows_error_count:
        raise RuntimeError("Row errors require reconciliation before source handoff")
    if status.status_code != "SUCCESS":
        raise streaming.StreamingIngestError(
            streaming.StreamingIngestErrorCode.INVALID_CHANNEL_ERROR,
            status.status_code,
            409,
            "Conflict",
        )

    committed = min(submitted, parse_offset(status.latest_committed_offset_token))
    if committed > source.committed:
        source.acknowledge(committed)


def parse_offset(token: Optional[str]) -> int:
    """Decode this sample's numeric offset; application tokens may be opaque."""
    return 0 if token is None else int(token)


# Channel lifecycle

class NamedProducer:
    """Own one stable named channel and its client."""

    def __init__(self, factory=create_client):
        self.factory = factory
        self.client = None
        self.channel = None

    def open(self) -> int:
        if self.client is None:
            self.client = self.factory()
        self.channel, status = self.client.open_channel(CHANNEL_NAME)
        if status.rows_error_count:
            raise RuntimeError("Row errors require reconciliation before source handoff")
        return parse_offset(status.latest_committed_offset_token)

    def recover(self, error: streaming.StreamingIngestError) -> int:
        """Reopen the channel without replacing its committed offset."""
        if error.error_code.value == "InvalidClientError":
            self.close(False)
        elif self.channel is not None:
            try:
                # Close only the local handle. Do not drop the named channel or
                # replace its server-side committed offset.
                self.channel.close(wait_for_flush=False, timeout_seconds=0)
            except streaming.StreamingIngestError:
                pass

        try:
            return self.open()
        except streaming.StreamingIngestError as reopened:
            if reopened.error_code.value not in {
                "InvalidClientError",
                "ClosedClientError",
            }:
                raise
            self.close(False)
            return self.open()

    def close(self, flush: bool) -> None:
        if self.client is None:
            return
        try:
            self.client.close(wait_for_flush=flush, timeout_seconds=30)
        finally:
            self.client = None
            self.channel = None


if __name__ == "__main__":
    main()
